core/generated-sources/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/json.kt

```diff
@@ -859,42 +867,103 @@ private val valueTypes =
 
 internal fun KlaxonJson.encodeRow(frame: ColumnsContainer<*>, index: Int): JsonObject? {
     val values = frame.columns().map { col ->
-        when {
-            col is ColumnGroup<*> -> encodeRow(col, index)
-            col is FrameColumn<*> -> encodeFrame(col[index])
-            col.isList() -> {
-                col[index]?.let { array(it as List<*>) } ?: array()
-            }
+        when (col) {
+            is ColumnGroup<*> -> encodeRow(col, index)
+            is FrameColumn<*> -> encodeFrame(col[index])
+            else -> encodePrimitiveData(col, index)
+        }.let { col.name to it }
+    }
+    if (values.isEmpty()) return null
+    return obj(values)
+}
 
-            col.typeClass in valueTypes -> {
-                val v = col[index]
-                if ((v is Double && v.isNaN()) || (v is Float && v.isNaN())) {
-                    v.toString()
-                } else v
+internal object SerializationKeys {
+    const val DATA = "data"
+    const val METADATA = "metadata"
+    const val KIND = "kind"
+    const val NCOL = "ncol"
+    const val NROW = "nrow"
+    const val VERSION = "\$version"
+    const val COLUMNS = "columns"
+    const val KOTLIN_DATAFRAME = "kotlin_dataframe"
+}
+
+internal const val SERIALIZATION_VERSION = "2.0.0"
+
+internal fun KlaxonJson.encodeRowWithMetadata(
+    frame: ColumnsContainer<*>,
+    index: Int,
+    rowLimit: Int? = null
+): JsonObject? {
+    val values = frame.columns().map { col ->
+        when (col) {
+            is ColumnGroup<*> -> obj(
+                DATA to encodeRowWithMetadata(col, index, rowLimit),
+                METADATA to obj(KIND to ColumnKind.Group.toString())
+            )
+
+            is FrameColumn<*> -> {
+                val data = if (rowLimit == null) encodeFrameWithMetadata(col[index])
+                else encodeFrameWithMetadata(col[index].rows().take(rowLimit).toDataFrame(), rowLimit)
```

Comment:
I believe rows().take(rowLimit).toDataFrame() can be simplified to `take(rowLimit)`. Can you check if result is the same?

Reply:
I checked, but it doesn't work without an explicit cast as `take(rowLimit)` returns List<DataRow>

Reply:
`else encodeFrameWithMetadata(col[index].take(rowLimit), rowLimit)`
Like this

Reply:
and there's similar place in `internal fun KlaxonJson.encodeDataFrameWithMetadata`

Reply:
`frame.rows().take(rowLimit).toDataFrame()` -> `frame.take(rowLimit)`

Reply:
Oh, now I got it! Thanks!